src/control/lib/netdetect/netdetect.go

```diff
@@ -788,6 +817,58 @@ func ValidateProviderConfig(device string, provider string) error {
 			continue
 		}
 		deviceScanCfg.targetDevice = C.GoString(fi.domain_attr.name)
+
+		// Implements a workaround to handle the current psm2 provider behavior
+		// that reports fi.domain_attr.name as "psm2" instead of an actual device
+		// name like "hfi1_0".
+		//
+		// The work around behaves in two different ways depending on the data available.
+		// When the value of fi.src_addr.unit is 0 or greater, it is used to explicitly
+		// define an hfi1_ unit by appending that number to the "hfi1_" string.
+		//
+		// When the value of fi.src_addr.unit is -1, it means that all hfi1_x units in the system
+		// are supported by the psm2 provider.  In this case, we need to know how many hfi1_x units
+		// there are, and can look for a provider/device match against each item in the list.
+		//
+		// Note that "hfi1_x" is converted to a system device name such as "ibx"
+		// by the call to GetAffinityForDevice()
+		if C.GoString(fi.fabric_attr.prov_name) == "psm2" {
+			if strings.Contains(deviceScanCfg.targetDevice, "psm2") {
+				log.Debugf("psm2 provider and psm2 device found.")
+				hfiUnit := C.getHFIUnit(fi.src_addr)
+				switch hfiUnit {
+				case allHFIUsed:
+					for deviceID := 0; deviceID < hfiDeviceCount; deviceID++ {
+						deviceScanCfg.targetDevice = fmt.Sprintf("hfi1_%d", deviceID)
+						deviceAffinity, err := GetAffinityForDevice(deviceScanCfg)
+						if err != nil {
+							continue
+						}
+						if deviceAffinity.DeviceName == device {
+							log.Debugf("Device %s supports provider: %s", device, provider)
+							return nil
+						}
+					}
+					continue
+				case badAddress:
+					log.Debugf("The fi.src_addr was null - not adding hfi1_n scan result")
+					continue
+				default: // The hfi unit is stated explicitly
+					log.Debugf("Found hfi1_%d (actual)", hfiUnit)
+					deviceScanCfg.targetDevice = fmt.Sprintf("hfi1_%d", hfiUnit)
+					deviceAffinity, err := GetAffinityForDevice(deviceScanCfg)
+					if err != nil {
+						continue
+					}
+					if deviceAffinity.DeviceName == device {
+						log.Debugf("Device %s supports provider: %s", device, provider)
+						return nil
+					}
```

Comment:
Can this chunk be shared? I.e. a helper method of some kind?

Reply:
I've consolidated it.